subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java

```diff
@@ -434,7 +434,16 @@ private static class ClassBuilderImpl extends ClassVisitorScope implements Class
         private static final Type DEFAULT_PROPERTY_TYPE = getType(DefaultProperty.class);
         private static final Type BUILD_SERVICE_PROVIDER_TYPE = getType("Lorg/gradle/api/services/internal/BuildServiceProvider;");
         private static final Type INSTRUMENTED_EXECUTION_ACCESS_TYPE = getType("Lorg/gradle/internal/classpath/InstrumentedExecutionAccess;");
-        private static final Set<Type> DISALLOWED_AT_EXECUTION_INJECTED_SERVICES_TYPES = ImmutableSet.of(getType(Project.class), getType(Gradle.class));
+
+        // This set is unlikely will be changing oftenly, so instead of introducing additional level of indirection
+        // we are storing it here despite it's CC-related logic.
+        // Full prohibited hierarchy stored since there is no way to efficiently check hierarchy of `org.objectweb.asm.Type`
```

Comment:
```suggestion
        // The full prohibited hierarchy is stored because there is no efficient way to check the class hierarchy via `org.objectweb.asm.Type`.
```